src/Service/EolValidator.php

```diff
@@ -0,0 +1,218 @@
+                $errors[] = $validationResult;
+            }
+        }
+
+        return $errors;
+    }
+
+    /**
+     * Validates a given service and version.
+     *
+     * @param string $serviceName
+     * @param string $serviceVersion
+     * @param int $errorLevel
+     * @return string
+     */
+    public function validateService(string $serviceName, string $serviceVersion, int $errorLevel) : string
+    {
+        // Get the EOL configurations for the current service.
+        $serviceConfigs = $this->getServiceEolConfigs($serviceName);
+
+        // Check if configurations exist for the current service and version.
+        $versionConfigs = array_filter($serviceConfigs, function ($v) use ($serviceVersion) {
+            return Semver::satisfies($serviceVersion, sprintf('%s.x', $v['version']));
+        });
+
+        // If there are no configurations found for the current service and version,
+        // or if an EOL is not defined or is invalid, return a message with details.
+        if (!$versionConfigs || $versionConfigs[array_key_first($versionConfigs)]['eol'] === null) {
+            return $errorLevel === ValidatorInterface::LEVEL_WARNING ? sprintf(
+                'Unknown or invalid EOL defined for %s %s',
+                $serviceName,
+                $serviceVersion
+            ) : '';
+        }
+
+        $eolDate = Carbon::createFromTimestamp($versionConfigs[array_key_first($versionConfigs)]['eol']);
+
+        // If the EOL is in the past, issue a warning.
+        // If the EOL is in the future, but within a three month period, issue a notice.
+        return $this->getServiceEolNotifications($eolDate, $errorLevel, $serviceName, $serviceVersion);
+    }
+
+    /**
+     * Gets the version of a given service.
+     *
+     * @param string $serviceName
+     * @return string
+     * @throws ServiceMismatchException
+     */
+    private function getServiceVersion(string $serviceName) : string
+    {
+        switch ($serviceName) {
+            case ServiceInterface::NAME_PHP:
+                $serviceVersion = PHP_VERSION;
+                break;
+            case ServiceInterface::NAME_ELASTICSEARCH:
+                $serviceVersion = $this->elasticSearch->getVersion();
+                break;
+            default:
+                $service = $this->serviceFactory->create($serviceName);
+                $serviceVersion = $service->getVersion();
+                break;
+        }
+
+        return $serviceVersion;
+    }
+
+    /**
+     * Gets the EOL configurations for the current service from eol.yaml.
+     *
+     * @param string $serviceName
+     * @return array
+     */
+    private function getServiceEolConfigs(string $serviceName) : array
+    {
+        // Check if the the configuration yaml file exists, and retrieve it's path.
+        if (file_exists($this->fileList->getServiceEolsConfig())) {
+            $configs = Yaml::parseFile($this->fileList->getServiceEolsConfig());
```

Comment:
It's better to cache the result of `Yaml::parse` to avoid reading a config file for each service.